arch/lkl/kernel/syscalls.c

```diff
@@ -295,6 +316,21 @@ static int idle_host_task_loop(void *unused)
 	for (;;) {
 		lkl_cpu_put();
 		lkl_ops->sem_down(ti->sched_sem);
+
+		if (lkl_shutdown) {
+
+			/**
+			 * Notify the host of the shutdown.
+			 *
+			 * Note that we are not releasing the CPU lock here, which allows
+			 * the termination code to do syscalls directly.
```

Comment:
I don't understand this comment.  Any thread can call `lkl_syscall`, this one happens to already own the CPU lock.  The `thread_sched_jb` call above will run through the scheduler and may wake up sleeping tasks (while transfering ownership of the CPU lock to them) and allow them to make system calls as well.

Reply:
The bigger problem here is that the termination calls should not come from the idle task. Apparently the kernel can become unhappy when substantial work is done in the idle task (e.g. sleeps or scheduler invocations). Therefore I see very rare kernel panics with this, complaining about the work in idle task.

We should establish a separate kernel task, which doesn't get killed and can execute the shutdown sequence. I propose that we should do this as part of a future PR, as this one already improves stability quite substantially, and we can build on top of it.